gax-grpc/src/main/java/com/google/api/gax/grpc/OperationPollingCallable.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2017, Google Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.grpc;
+
+import com.google.api.core.ApiClock;
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutures;
+import com.google.api.gax.retrying.ExponentialRetryAlgorithm;
+import com.google.api.gax.retrying.ResultRetryAlgorithm;
+import com.google.api.gax.retrying.RetrySettings;
+import com.google.api.gax.retrying.TimedAttemptSettings;
+import com.google.longrunning.GetOperationRequest;
+import com.google.longrunning.Operation;
+import com.google.longrunning.OperationsClient;
+import io.grpc.Status;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * A future callable, which relies on callback chaining to execute server polling (asking for a
+ * status of a specific operation: in progress, completed, canceled etc.). This implementation is
+ * called from {@link AttemptCallable}, essentially using for polling the same logic which is used
+ * for retrying.
+ *
+ * @param <RequestT> type of the request
+ */
+class OperationPollingCallable<RequestT> implements FutureCallable<RequestT, Operation> {
```

Comment:
It does the polling, not by having an explicit loop inside, but by returning "correct" futures according to conditions. 

(I do not claim the following description to be completely correct or rigorous, but I believe it represent a valid point of view on this case):

This class is closer to functional programming approach than it is usual in java. I don't mean it is actually functional, especially since this class has state, but overall it is still somewhat close in essence. 

Specifically it does not have a procedural "while" loop for polling or anything like that which states how to perform the polling. 

Instead, this class contains two stateless functional classes `OperationRetryAlgorithm` and `OperationTimedAlgorithm` (aka first-class functions (or sets of functions), like in functional programming), which, when consumed by the "outer world" (`com.google.api.gax.retrying` package and `RetryingCallable`) make retrying to perform polling. 

I.e. one can put functions (methods in functional classes), declared in this class as arguments to the "retrying function", which will transform it into polling. The actual polling is done by retrying logic.